core/src/main/java/com/rudderstack/android/sdk/core/RudderDeviceModeManager.java

```diff
@@ -221,40 +222,51 @@ private void isDeviceModeFactoriesNotPresent() {
      *
      */
     private void replayMessageQueue() {
-        List<Integer> messageIds = new ArrayList<>();
-        List<String> messages = new ArrayList<>();
-        do {
-            dbPersistentManager.fetchDeviceModeEventsFromDb(messageIds, messages, MAX_FLUSH_QUEUE_SIZE);
-            RudderLogger.logDebug(String.format(Locale.US, "RudderDeviceModeManager: replayMessageQueue: replaying old messages with factories. Count: %d", messageIds.size()));
-            for (int i = 0; i < messageIds.size(); i++) {
-                try {
-                    RudderMessage message = gson.fromJson(messages.get(i), RudderMessage.class);
-                    makeFactoryDump(message, messageIds.get(i), true);
-                } catch (Exception e) {
-                    RudderLogger.logError(String.format(Locale.US, "RudderDeviceModeManager: replayMessageQueue: Exception in dumping message %s due to %s", messages.get(i), e.getMessage()));
+        synchronized (replayMessageQueueLock) {
+            List<Integer> messageIds = new ArrayList<>();
+            List<String> messages = new ArrayList<>();
+            int offset = 0;
```

Comment:
Wouldn't the events table get updated while the replayMessageQueue() is getting executed, and then un-processed events would be shifted before offset right ?

Reply:
Correct, I would improve the code. Thanks for pointing this out.
Improvement: Will call this offset DB API only in case DMT is active, else I'll call the normal API.